File: python/linked_list/reverse_between.py
class ListNode:
    def __init__(self, val: int = 0, next: None = None) -> None:
        self.val = val
        self.next = next

class Solution:
    def reverseBetween(self, head: ListNode, left: int,  right: int) -> ListNode:
        if not head or left == right:
            return head

        dummy = ListNode(0)
        dummy.next = head
        prev = dummy

        for _ in range(left - 1):
            prev = prev.next


        reverse_start = prev.next
        then = reverse_start.next

        for _ in range(right - left):
            reverse_start.next = then.next
            then.next = prev.next
            prev.next = then
            then = reverse_start.next

        return dummy.next

File: golang/linked_list/reverse_between.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseBetween(head *ListNode, left int, right int) *ListNode {
	if head == nil || left == right {
		return head
	}

	dummy := &ListNode{Next: head}

	prev := dummy

	for i := 0; i < left-1; i++ {
		prev = prev.Next
	}

	reverseStart := prev.Next
	then := reverseStart.Next

	for i := 0; i < right-left; i++ {
		reverseStart.Next = then.Next

		then.Next = prev.Next
		prev.Next = then
		then = reverseStart.Next
	}

	return dummy.Next
}
